### ISSUE TITLE
Editor doesn't reload changed resources.

### ISSUE THREAD

**Issue by Geequlim** [godotengine/godot#4769]
**Operating system or device - Godot version:**
Arch Linux 64bit - Godot 2.1 alpha compiled with b2a4908e9cc802a838a67f92dd66a17d3894c619

**Issue description** (what happened, and what was expected):

The editor doesn't reload changed resources generated by ResourceSaver.

**Steps to reproduce:**
1. Create a resource and save it with ResourceSaver in script (`version 1`).
2. Open the resource with editor it seems everything is fine.
3. Create another different resource and save it to the same path (`version 2`).
4. Open the overwrited resource you can see it still be the `version 1`.
5. Check the resource file with other text editor, it is correct.
6. Reopen the project and open the resource file , it will show `version 2`

**Link to minimal example project** (optional but very welcome):

https://drive.google.com/file/d/0B0AolAabr5edbXpEZkVnVXRXRTA/view

**Comment by jjay** [comment]
@Geequlim pointed this issue when the editor does not respond to filesystem changes. I encountered the same problem, but the editor does not respond to actions prepared in the editor  (if I create a scene from a node and save it to res://existed_scene.tscn, existed_scene.tscn will not be updated in tab until i reopen it).

I made some tests to learn how the editor reacts to different changes, when `Resource.changed` signal emits and how the editor saves resources itself. Here are some results:

* The editor reacts to some filesystem changes in different ways based on the type of changes:
  * It offers to reimport Resources if the Sources changed
  * It invalidates FileDialogs if file/dir added/removed inside res://
  * There is no reaction if Resource file changed
* Some resources emits `Resource.changed` signal  and it is not related to filesystem changes
* There are a lot of places where `ResourceSaver.save` is used, and none of them generate any signals, so the editor docks keep working with the old versions of resources.
* EditorFileSystem makes expensive scan and does't rely on any os-level notifications like `inotify` or `kqueue`.

Looks like there is a missed signal somewhere. Like `Resource.file_changed`. It would be cool if the editor (and extensions) could rely on that signal and give to the user some positive feedback about what is going on.

I fell in love with Godot, and I try to integrate it in company I am working for. So I'm ready to start working on this problem. Point me please if I've missed something important.

**Comment by reduz** [comment]
I don' t really understand the issue, but here's what I can help with so far:

1) ResourceSaver::save has a callback that EditorFilesystem uses to tell when a file changed and update it
2) If you save a resource and overwrite an existing loaded resource, the existing loaded resource will lose it's reference to the file but it will keep existing. There isn't any kind of merging going on and this might result in unexpected stuff. Not sure what the ideal would be on this situation.
3) The only reason inotify/kqueue is not used is portability (there is nothing like this on other OSs). The scan is not really that expensive though, as a lot of information is cached

**Comment by jjay** [comment]
Thank you @reduz for the fast explanation! I do not consider that save_callback at all. I will go learn a bit more about ResourceSaver and Resource ecosystem =)

**Comment by reduz** [comment]
So, what do we do with this one?

**Comment by jjay** [comment]
Well, this issue a little bit confusing, because it is not related to ResourceSaver actually. In example project, you need to run the game, and press button. The ResourceSaver changes some scene and writes it to project dir. When you stop the game, editor won't show that changes, because it is not watching for filesystem changes.

#4820 do nothing with ResourceSaver, but a workaround with editor. Editor checks current opened scene for changes on getting focus, and reload that scene from disk if it was modified. Editor also asks user what to do if scene has local modifications.

**Comment by reduz** [comment]
Recently we made it so Godot reloads resources that changed on disk, so this should be closed

**Comment by fancyn3rd** [comment]
Why this is closed? Its still not working in godot 2.1.4 :(
(everything of the first post applies)

**Comment by khoulihan** [comment]
This still seems to be a problem in 3.0.2. My workaround was to reload the resource immediately after saving it.

**Comment by firecloud888** [comment]
Still a problem in 3.1, and I cannot find a way to reopen the editor tab manually in code.

**Comment by Torguen** [comment]
It remains a problem in 3.2 beta 1 monkey.
It takes a long time to update a folder in which you have placed a graphic.

I created a graphic with Aseprite and then saved it in the project folder.
But Godot did not update the contents of the folder and the graphic did not appear, after doing some things and displaying the editor again, then Godot updated it and now I can see the graphic.

**Comment by RabbitB** [comment]
As I stated in #30302, this is still an issue in Godot 3.2.1

**Comment by amoriqbal** [comment]
We need a reload() function besides the load() function. Its clearly stated in the docs of 3.2 stable, that resources are loaded only once and stores it in main memory.  We need a reload() function, that deletes the instance of the resource from memory, and calls load() function again, as if it is the first time that the resource is being loaded.
I am facing this problem that my godot application, after export, is not reloading the resource even after the resource has actually been changed on disk (by an external script).

**Comment by Calinou** [comment]
@amoriqbal Please open a [proposal](https://github.com/godotengine/godot-proposals) to request new features.

**Comment by Shadowblitz16** [comment]
still happening on 3.2.3

**Comment by SubSage** [comment]
Confirming still an issue, I'd like to suggest re-opening this. @akien-mga

**Comment by akien-mga** [comment]
@SubSage This issue was closed 4 years ago. The new issue is #30302 as mentioned above.
